### PR TITLE
Add additional tags to ASP.NET Core metrics

### LINKED ISSUE #3083: Improve ASP.NET Core metrics instrumentation attribute compliance
Today our AspNetCore metrics instrumentation captures a small subset...

https://github.com/open-telemetry/opentelemetry-dotnet/blob/96b043e35fcf407147a9ad9c4fae2b6915f3a9ab/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs#L58-L61

...of the attributes defined in the [spec semantic conventions](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#attributes). Notably `http.target` (route template) and `http.host` are missing which are probably high-value/nice-to-have.

If anyone would like to take this on I am available to provide guidance.

Relates to #1269

### PR DESCRIPTION
**Pull request by vishweshbankwar**
Fixes #3083 

## Changes

Please provide a brief description of the changes here.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by CodeBlanch — APPROVED** [review]
LGTM

Just checking what Prometheus .NET client does:

* https://github.com/prometheus-net/prometheus-net/blob/ff337560e66af5dd45070014d6e554febb599f16/Prometheus.AspNetCore/HttpMetrics/HttpRequestMiddlewareBase.cs#L264-L277
* https://github.com/prometheus-net/prometheus-net/blob/ff337560e66af5dd45070014d6e554febb599f16/Prometheus.AspNetCore/HttpMetrics/HttpRequestMiddlewareBase.cs#L195

Seems to allow users to map known route values as dimensions. We may want to do similar at some point.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 1]
```diff
@@ -51,16 +54,58 @@ public override void OnStopActivity(Activity activity, object payload)
...
+                host = context.Request.Host.Host;
+            }
+            else
+            {
+                host = context.Request.Host.Host + ":" + context.Request.Host.Port;
+            }
+
+            TagList tags;
+
+#if NETCOREAPP
```
Comment (alanwest) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:70]: We have a `net6.0` and `netstandard2.1` build of this instrumentation. Does this directive apply to those targets?

Comment (alanwest) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:70]: Actually, now I'm wondering why we need a `netstandard2.1` target given that we have a `netcoreapp3.1` target.

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:70]: This directive only applies to `netcoreapp3.1` and `net6.0`. It does not apply to `netstandard2.1`\`netstandard2.0`

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:70]: could you add a comment why this compiler directive is relevant. ?

**Review by alanwest — APPROVED** [review]
LGTM. Still curious if we need the `netstandard2.x` targets anymore, but that's unrelated to this PR.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs (hunk 1 above)
Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:72]: @cijothomas - Added - Please review.

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:72]: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInListener.cs#L243 -- could you check this part.. it does use reflection to get the releveant info on routes...

Comment (vishweshbankwar) [reply src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:72]: We do not have that information available during OnStop event.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 2]
```diff
@@ -51,16 +54,60 @@ public override void OnStopActivity(Activity activity, object payload)
...
             {
-                { SemanticConventions.AttributeHttpMethod, context.Request.Method },
+                tags = new TagList
+                {
+                    { SemanticConventions.AttributeHttpFlavor, context.Request.Protocol },
+                    { SemanticConventions.AttributeHttpScheme, context.Request.Scheme },
+                    { SemanticConventions.AttributeHttpMethod, context.Request.Method },
+                    { SemanticConventions.AttributeHttpHost, host },
+                    { SemanticConventions.AttributeHttpTarget, target },
+                    { SemanticConventions.AttributeHttpStatusCode, context.Response.StatusCode },
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:86]: nit: Spec says this should be string, but I think `StatusCode` is an int here.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 3]
```diff
@@ -51,16 +54,60 @@ public override void OnStopActivity(Activity activity, object payload)
...
 
+            string host;
+
+            if (context.Request.Host.Port is null or 80 or 443)
+            {
+                host = context.Request.Host.Host;
+            }
+            else
+            {
+                host = context.Request.Host.Host + ":" + context.Request.Host.Port;
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:65]: for a follow up - see if we avoid this string alloc, by using cache/other technique?

**Review by cijothomas — APPROVED** [review]
LGTM left couple of nits.
Please add to changelog.md as this is a noteworthy use-facing change.

File: src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs [hunk 4]
```diff
@@ -51,16 +54,60 @@ public override void OnStopActivity(Activity activity, object payload)
...
+            tags = new TagList
             {
-                { SemanticConventions.AttributeHttpMethod, context.Request.Method },
-                { SemanticConventions.AttributeHttpScheme, context.Request.Scheme },
-                { SemanticConventions.AttributeHttpStatusCode, context.Response.StatusCode },
                 { SemanticConventions.AttributeHttpFlavor, context.Request.Protocol },
+                { SemanticConventions.AttributeHttpScheme, context.Request.Scheme },
+                { SemanticConventions.AttributeHttpMethod, context.Request.Method },
+                { SemanticConventions.AttributeHttpHost, host },
+                { SemanticConventions.AttributeHttpStatusCode, context.Response.StatusCode.ToString() },
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.AspNetCore/Implementation/HttpInMetricsListener.cs:107]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3247/files#r866970771 same comment. we can follow up to fix the string alloc.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**